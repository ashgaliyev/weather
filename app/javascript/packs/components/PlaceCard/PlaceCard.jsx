import PropTypes from "prop-types";
import React from "react";
import { SettingsContext } from "../SettingsProvider";
import placeholder from "./icon_placeholder.png";
import { weatherShape } from "../../utils/types";

function PlaceLink({ children, onClick }) {
  if (onClick) {
    return (
      <div
        className="bg-white rounded-lg shadow-lg flex flex-row p-4 pl-0 max-w-sm hover:cursor-pointer"
        onClick={onClick}
        onKeyDown={onClick}
        role="button"
        tabIndex={0}
        data-test-id="place-card"
      >
        {children}
      </div>
    );
  }

  return (
    <div className="bg-white rounded-lg shadow-lg flex flex-row p-4 pl-0 max-w-sm">
      {children}
    </div>
  );
}

PlaceLink.propTypes = {
  children: PropTypes.node.isRequired,
  onClick: PropTypes.func,
};

PlaceLink.defaultProps = {
  onClick: null,
};

function PlaceCard({ heading, weather, onClick }) {
  if (!weather) {
    return (
      <PlaceLink onClick={onClick}>
        <div className="flex-none w-24">
          <img src={placeholder} alt="No data" />
        </div>
        <div className="flex-1 w-32">
          <div className="text-2xl">{heading.slice(0, 20)}</div>
          <div className="text-xs text-right text-slate-400">Updating...</div>
        </div>
      </PlaceLink>
    );
  }

  const { tempMin, tempMax, windSpeed, icon, description } = weather;

  const { tempType } = React.useContext(SettingsContext);

  const tempMinC = Math.round(tempMin - 273.15);
  const tempMaxC = Math.round(tempMax - 273.15);

  const tempMinF = Math.round((tempMin - 273.15) * (9 / 5) + 32);
  const tempMaxF = Math.round((tempMax - 273.15) * (9 / 5) + 32);

  const tempRange =
    tempType === "Celsius"
      ? `${tempMinC}째C..${tempMaxC}째C`
      : `${tempMinF}째F..${tempMaxF}째F`;

  return (
    <PlaceLink onClick={onClick}>
      <div className="flex-none w-24">
        <img
          src={`https://openweathermap.org/img/wn/${icon}@2x.png`}
          alt={description}
        />
      </div>
      <div className="flex-1 w-32">
        <div className="text-2xl">{heading.slice(0, 20)}</div>
        <span>{tempRange}</span>
        <div>{windSpeed} m/s</div>
      </div>
    </PlaceLink>
  );
}

PlaceCard.propTypes = {
  heading: PropTypes.string.isRequired,
  onClick: PropTypes.func,
  weather: weatherShape,
};

PlaceCard.defaultProps = {
  weather: null,
  onClick: null,
};

export default PlaceCard;
