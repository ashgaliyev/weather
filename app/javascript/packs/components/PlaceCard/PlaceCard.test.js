import React from "react";
import renderer from "react-test-renderer";
import PlaceCard from "./PlaceCard";

describe("PlaceCard", () => {
  beforeEach(() => {
    React.useContext = jest.fn().mockReturnValue({ tempType: "c" });
  });

  describe("when place has no forecast", () => {
    it("renders correctly", () => {
      const props = {
        heading: "Paris",
        onClick: jest.fn(),
      };

      const tree = renderer.create(<PlaceCard {...props} />).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe("when place has forecast", () => {
    it("renders correctly", () => {
      const props = {
        heading: "Paris",
        onClick: jest.fn(),
        weather: {
          tempMin: 300,
          tempMax: 310,
          icon: "01d",
          description: "clear sky",
          windSpeed: 5,
          date: "2023-05-01",
        },
      };

      const tree = renderer.create(<PlaceCard {...props} />).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});
