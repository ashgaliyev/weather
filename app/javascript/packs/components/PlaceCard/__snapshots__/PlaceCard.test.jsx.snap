// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlaceCard when place has forecast renders correctly 1`] = `
<div
  className="bg-white rounded-lg shadow-lg flex flex-row p-4 pl-0 max-w-sm hover:cursor-pointer"
  data-test-id="place-card"
  onClick={[MockFunction]}
  onKeyDown={[MockFunction]}
  role="button"
  tabIndex={0}
>
  <div
    className="flex-none w-24"
  >
    <img
      alt="clear sky"
      src="https://openweathermap.org/img/wn/01d@2x.png"
    />
  </div>
  <div
    className="flex-1 w-32"
  >
    <div
      className="text-2xl"
    >
      Paris
    </div>
    <span>
      80°F..98°F
    </span>
    <div>
      5
       m/s
    </div>
  </div>
</div>
`;

exports[`PlaceCard when place has no forecast renders correctly 1`] = `
<div
  className="bg-white rounded-lg shadow-lg flex flex-row p-4 pl-0 max-w-sm hover:cursor-pointer"
  data-test-id="place-card"
  onClick={[MockFunction]}
  onKeyDown={[MockFunction]}
  role="button"
  tabIndex={0}
>
  <div
    className="flex-none w-24"
  >
    <img
      alt="No data"
      src={{}}
    />
  </div>
  <div
    className="flex-1 w-32"
  >
    <div
      className="text-2xl"
    >
      Paris
    </div>
    <div
      className="text-xs text-right text-slate-400"
    >
      Updating...
    </div>
  </div>
</div>
`;
